### PR TITLE
Loop unroll; avoid bounds checks

### PR DESCRIPTION
**Pull request by benaadams**

### CODE REVIEW

**Comment by davidfowl** [comment]
@benaadams Before after numbers?

**Comment by benaadams** [comment]
Will benchmark

File: test/Microsoft.AspNet.Server.KestrelTests/AsciiDecoder.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public class AsciiDecoderTests
         [Fact]
         private void FullByteRangeSupported()
         {
-            var byteRange = Enumerable.Range(0, 255).Select(x => (byte)x).ToArray();
+            var byteRange = Enumerable.Range(0, 256).Select(x => (byte)x).ToArray();
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Server.KestrelTests/AsciiDecoder.cs]: ?

Comment (benaadams) [inline test/Microsoft.AspNet.Server.KestrelTests/AsciiDecoder.cs]: I got it wrong, is count not upper bound. Tests still pass either way.

**Comment by benaadams** [comment]
Turns out the cost of fixed is too high vs straight array.

```ini
BenchmarkDotNet=v0.7.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i5-3470S CPU @ 2.90GHz, ProcessorCount=4
HostCLR=MS.NET 4.0.30319.42000, Arch=32-bit
Type=AsciiToString  Mode=Throughput  .NET=HostFramework
```

                  Method | Platform |       Jit |   AvrTime |    StdDev |           op/s |
------------------------ |--------- |---------- |---------- |---------- |--------------- |
         SequentialArray |      X64 | LegacyJit | 1.5266 ns | 0.0029 ns | 655,062,947.42 |
       SequentialPointer |      X64 | LegacyJit | 1.9622 ns | 0.0277 ns | 509,629,020.56 |
 UnrolledParallelPointer |      X64 | LegacyJit | 2.1806 ns | 0.0645 ns | 458,579,769.91 |
         UnrolledPointer |      X64 | LegacyJit | 2.1801 ns | 0.0276 ns | 458,692,301.62 |
         SequentialArray |      X64 |    RyuJit | 1.4125 ns | 0.0224 ns | 707,981,977.98 |
       SequentialPointer |      X64 |    RyuJit | 1.6971 ns | 0.0238 ns | 589,249,356.03 |
 UnrolledParallelPointer |      X64 |    RyuJit | 1.8363 ns | 0.0261 ns | 544,580,446.84 |
         UnrolledPointer |      X64 |    RyuJit | 1.8825 ns | 0.0235 ns | 531,199,310.82 |
         SequentialArray |      X86 | LegacyJit | 1.2736 ns | 0.0173 ns | 785,193,458.26 |
       SequentialPointer |      X86 | LegacyJit | 1.5600 ns | 0.0227 ns | 641,014,201.10 |
 UnrolledParallelPointer |      X86 | LegacyJit | 1.5364 ns | 0.0208 ns | 650,892,834.97 |
         UnrolledPointer |      X86 | LegacyJit | 1.5262 ns | 0.0598 ns | 655,234,541.43 |

Closing this

**Comment by benaadams** [comment]
Test code: https://gist.github.com/benaadams/43407df97c69a4ae86e8

**Comment by benaadams** [comment]
Retesting this was will a length always of 0

**Comment by benaadams** [comment]
Fixed test cases https://gist.github.com/benaadams/079a1ded2e14bc243fd5

**Comment by benaadams** [comment]
```ini
BenchmarkDotNet=v0.7.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz, ProcessorCount=8
HostCLR=MS.NET 4.0.30319.42000, Arch=64-bit  [RyuJIT]
Type=AsciiToString  Mode=Throughput  Platform=X64  Jit=RyuJit  .NET=HostFramework
```

                  Method | Param |       AvrTime |     StdDev |           op/s |
------------------------ |------ |-------------- |----------- |--------------- |
         SequentialArray |     1 |     1.8446 ns |  0.0359 ns | 542,130,112.58 |
         SequentialArray |     2 |     1.8347 ns |  0.0301 ns | 545,041,308.79 |
         SequentialArray |     8 |     2.6034 ns |  0.0473 ns | 384,119,102.12 |
         SequentialArray |    64 |    15.5845 ns |  1.2281 ns |  64,166,181.67 |
         SequentialArray |   512 |   109.1622 ns |  3.9246 ns |   9,160,708.35 |
         SequentialArray |  1024 |   216.8628 ns |  4.3980 ns |   4,611,210.56 |
         SequentialArray |  4096 |   838.0113 ns |  8.0319 ns |   1,193,301.35 |
         SequentialArray |  8192 | 1,665.9055 ns | 17.9379 ns |     600,274.20 |
         SequentialArray | 16384 | 3,331.0892 ns | 31.2668 ns |     300,202.12 |
       SequentialPointer |     1 |     1.4709 ns |  0.0290 ns | 679,872,918.30 |
       SequentialPointer |     2 |     1.7573 ns |  0.0055 ns | 569,058,870.74 |
       SequentialPointer |     8 |     2.7821 ns |  0.0495 ns | 359,435,697.96 |
       SequentialPointer |    64 |    12.2456 ns |  0.2320 ns |  81,662,055.86 |
       SequentialPointer |   512 |    77.9059 ns |  1.3984 ns |  12,836,002.73 |
       SequentialPointer |  1024 |   147.6454 ns |  2.6145 ns |   6,772,985.46 |
       SequentialPointer |  4096 |   603.3346 ns | 10.7317 ns |   1,657,455.03 |
       SequentialPointer |  8192 | 1,203.2110 ns | 20.8846 ns |     831,109.43 |
       SequentialPointer | 16384 | 2,404.2652 ns |  1.2482 ns |     415,927.49 |
 UnrolledParallelPointer |     1 |     1.4456 ns |  0.0280 ns | 691,731,058.85 |
 UnrolledParallelPointer |     2 |     1.6387 ns |  0.0255 ns | 610,257,762.49 |
 UnrolledParallelPointer |     8 |     1.9083 ns |  0.0923 ns | 524,037,000.21 |
 UnrolledParallelPointer |    64 |     5.8002 ns |  0.1617 ns | 172,408,326.49 |
 UnrolledParallelPointer |   512 |    39.9602 ns |  1.6196 ns |  25,024,919.12 |
 UnrolledParallelPointer |  1024 |    77.5874 ns |  1.8344 ns |  12,888,690.04 |
 UnrolledParallelPointer |  4096 |   306.2025 ns |  4.9520 ns |   3,265,812.70 |
 UnrolledParallelPointer |  8192 |   606.6170 ns | 10.3170 ns |   1,648,486.66 |
 UnrolledParallelPointer | 16384 | 1,447.2493 ns | 25.6027 ns |     690,965.98 |
         UnrolledPointer |     1 |     1.5475 ns |  0.0068 ns | 646,184,034.29 |
         UnrolledPointer |     2 |     1.7582 ns |  0.0302 ns | 568,748,690.53 |
         UnrolledPointer |     8 |     2.3780 ns |  0.0419 ns | 420,515,824.62 |
         UnrolledPointer |    64 |     9.5331 ns |  0.1572 ns | 104,897,218.65 |
         UnrolledPointer |   512 |    68.6465 ns |  1.2514 ns |  14,567,379.93 |
         UnrolledPointer |  1024 |   134.3055 ns |  2.3314 ns |   7,445,713.61 |
         UnrolledPointer |  4096 |   528.5438 ns |  8.8499 ns |   1,891,990.90 |
         UnrolledPointer |  8192 | 1,054.4902 ns |  0.7537 ns |     948,325.53 |
         UnrolledPointer | 16384 | 2,105.9399 ns |  2.7216 ns |     474,847.36 |

**Comment by benaadams** [comment]
@davidfowl new implemenation is 
x 1.36 for 8 byte string
x 2 for 64 byte string
x 2.7 for 512 byte string

**Comment by benaadams** [comment]
Will come back to this

**Pull request closed without merging**